prov/sm2/src/sm2_progress.c

```diff
@@ -223,112 +144,79 @@ static int sm2_start_common(struct sm2_ep *ep, struct sm2_cmd *cmd,
 	}
 
+	sm2_get_peer_srx(ep)->owner_ops->free_entry(rx_entry);
+
+
 	return 0;
 }
 
 int sm2_unexp_start(struct fi_peer_rx_entry *rx_entry)
 {
-	struct sm2_cmd_ctx *cmd_ctx = rx_entry->peer_context;
+	struct sm2_fqe_ctx *fqe_ctx = rx_entry->peer_context;
 	int ret;
 
-	pthread_spin_lock(&cmd_ctx->ep->region->lock);
-	ret = sm2_start_common(cmd_ctx->ep, &cmd_ctx->cmd, rx_entry);
-	ofi_freestack_push(cmd_ctx->ep->cmd_ctx_fs, cmd_ctx);
-	pthread_spin_unlock(&cmd_ctx->ep->region->lock);
+	ret = sm2_start_common(fqe_ctx->ep, &fqe_ctx->fqe, rx_entry, false);
+	ofi_freestack_push(fqe_ctx->ep->fqe_ctx_fs, fqe_ctx);
 
 	return ret;
 }
 
-static void sm2_progress_connreq(struct sm2_ep *ep, struct sm2_cmd *cmd)
+static int sm2_alloc_fqe_ctx(struct sm2_ep *ep,
+		struct fi_peer_rx_entry *rx_entry, struct sm2_free_queue_entry *fqe)
 {
-	struct sm2_region *peer_smr;
-	struct sm2_inject_buf *tx_buf;
-	size_t inj_offset;
-	int64_t idx = -1;
-	int ret = 0;
-
-	inj_offset = (size_t) cmd->msg.hdr.src_data;
-	tx_buf = sm2_get_ptr(ep->region, inj_offset);
+	struct sm2_fqe_ctx *fqe_ctx;
 
-	ret = sm2_map_add(&sm2_prov, ep->region->map,
-			  (char *) tx_buf->data, &idx);
-	if (ret)
-		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
-			"Error processing mapping request\n");
-
-	peer_smr = sm2_peer_region(ep->region, idx);
-
-	if (peer_smr->pid != (int) cmd->msg.hdr.data) {
-		//TODO track and update/complete in error any transfers
-		//to or from old mapping
-		munmap(peer_smr, peer_smr->total_size);
-		sm2_map_to_region(&sm2_prov, ep->region->map, idx);
-		peer_smr = sm2_peer_region(ep->region, idx);
-	}
-	sm2_peer_data(peer_smr)[cmd->msg.hdr.id].addr.id = idx;
-	sm2_peer_data(ep->region)[idx].addr.id = cmd->msg.hdr.id;
-
-	smr_freestack_push(sm2_inject_pool(ep->region), tx_buf);
-	ofi_cirque_discard(sm2_cmd_queue(ep->region));
-	ep->region->cmd_cnt++;
-	assert(ep->region->map->num_peers > 0);
-	ep->region->max_sar_buf_per_peer = SM2_MAX_PEERS /
-		ep->region->map->num_peers;
-}
-
-static int sm2_alloc_cmd_ctx(struct sm2_ep *ep,
-		struct fi_peer_rx_entry *rx_entry, struct sm2_cmd *cmd)
-{
-	struct sm2_cmd_ctx *cmd_ctx;
-
-	if (ofi_freestack_isempty(ep->cmd_ctx_fs))
+	if (ofi_freestack_isempty(ep->fqe_ctx_fs))
 		return -FI_EAGAIN;
 
-	cmd_ctx = ofi_freestack_pop(ep->cmd_ctx_fs);
-	memcpy(&cmd_ctx->cmd, cmd, sizeof(*cmd));
-	cmd_ctx->ep = ep;
+	fqe_ctx = ofi_freestack_pop(ep->fqe_ctx_fs);
+	memcpy(&fqe_ctx->fqe, fqe, sizeof(*fqe));
+	fqe_ctx->ep = ep;
 
-	rx_entry->peer_context = cmd_ctx;
+	rx_entry->peer_context = fqe_ctx;
 
 	return FI_SUCCESS;
 }
 
-static int sm2_progress_cmd_msg(struct sm2_ep *ep, struct sm2_cmd *cmd)
+static int sm2_progress_recv_msg(struct sm2_ep *ep, struct sm2_free_queue_entry *fqe)
 {
 	struct fid_peer_srx *peer_srx = sm2_get_peer_srx(ep);
 	struct fi_peer_rx_entry *rx_entry;
 	fi_addr_t addr;
 	int ret;
 
-	addr = ep->region->map->peers[cmd->msg.hdr.id].fiaddr;
-	if (cmd->msg.hdr.op == ofi_op_tagged) {
+	addr = fqe->protocol_hdr.id;
```

Comment:
I think this actually will break the AV_USER_ID support too. Currently shm store the user id in `ep->region->map->peers[cmd->msg.hdr.id].fiaddr` so it can get the message from owner provider with correct addr ( @aingerson  correct me if I am wrong). AFAICT the `addr` here `fqe->protocol_hdr.id` is not the same thing.

Reply:
@shijin-aws I agree, I think they are not the same thing and that this is broken.